src/firmware/guest/types/snp.rs

```diff
@@ -643,238 +215,391 @@ impl TryFrom<&[u8]> for AttestationReportV2 {
+                cpuid_fam_id: Some(stepper.parse_bytes::<_, 0>()?),
+                cpuid_mod_id: Some(stepper.parse_bytes::<_, 0>()?),
+                cpuid_step: Some(stepper.parse_bytes::<_, 0>()?),
+                chip_id: stepper.parse_bytes::<_, 21>()?,
+                committed_tcb: TcbVersion::from_legacy_bytes(&stepper.parse_bytes::<[u8; 8], 0>()?),
+                current: stepper.parse_bytes::<_, 0>()?,
+                committed: stepper.parse_bytes::<_, 1>()?,
+                launch_tcb: TcbVersion::from_legacy_bytes(&stepper.parse_bytes::<[u8; 8], 1>()?),
+                signature: stepper.parse_bytes::<_, 168>()?,
+            },
+            ReportVariant::V4Turin => AttestationReport {
+                version: stepper.parse_bytes::<_, 0>()?,
+                guest_svn: stepper.parse_bytes::<_, 0>()?,
+                policy: stepper.parse_bytes::<_, 0>()?,
+                family_id: stepper.parse_bytes::<_, 0>()?,
+                image_id: stepper.parse_bytes::<_, 0>()?,
+                vmpl: stepper.parse_bytes::<_, 0>()?,
+                sig_algo: stepper.parse_bytes::<_, 0>()?,
+                current_tcb: TcbVersion::from_v4_bytes(&stepper.parse_bytes::<[u8; 8], 0>()?),
+                plat_info: stepper.parse_bytes::<_, 0>()?,
+                key_info: stepper.parse_bytes::<_, 0>()?,
+                report_data: stepper.parse_bytes::<_, 4>()?,
+                measurement: stepper.parse_bytes::<_, 0>()?,
+                host_data: stepper.parse_bytes::<_, 0>()?,
+                id_key_digest: stepper.parse_bytes::<_, 0>()?,
+                author_key_digest: stepper.parse_bytes::<_, 0>()?,
+                report_id: stepper.parse_bytes::<_, 0>()?,
+                report_id_ma: stepper.parse_bytes::<_, 0>()?,
+                reported_tcb: TcbVersion::from_v4_bytes(&stepper.parse_bytes::<[u8; 8], 0>()?),
+                cpuid_fam_id: Some(stepper.parse_bytes::<_, 0>()?),
+                cpuid_mod_id: Some(stepper.parse_bytes::<_, 0>()?),
+                cpuid_step: Some(stepper.parse_bytes::<_, 0>()?),
+                chip_id: stepper.parse_bytes::<_, 21>()?,
+                committed_tcb: TcbVersion::from_v4_bytes(&stepper.parse_bytes::<[u8; 8], 0>()?),
+                current: stepper.parse_bytes::<_, 0>()?,
+                committed: stepper.parse_bytes::<_, 1>()?,
+                launch_tcb: TcbVersion::from_v4_bytes(&stepper.parse_bytes::<[u8; 8], 1>()?),
+                signature: stepper.parse_bytes::<_, 168>()?,
+            },
+        })
+    }
+
+    /// Writes the Attestation Report back into the ASP binary format.
+    pub fn write_bytes(self, mut handle: impl Write) -> Result<(), std::io::Error> {
+        // Determine the variant based on version and CPUID step
+        let variant = if self.version == 2 {
+            ReportVariant::V2
+        } else if self.version == 3 || (self.version == 4 && self.cpuid_step.unwrap_or(0) < 0xB0) {
```

Comment:
```suggestion
        } else if self.version == 3 || (self.version == 4 && self.cpuid_fam_id.unwrap_or(0) < 0xB0) {
```

Should you be checking family here?